File: 2050-count-good-numbers/count-good-numbers.go
const MOD = int64(1e9 + 7);

func countGoodNumbers(n int64) int {
    evenCount := (n + 1) / 2;
    oddCount := n / 2;

    // 5 = {0, 2, 4, 6, 8}
    part1 := powMod(5, evenCount, MOD);
    // 4 = {2, 3, 5, 7}
    part2 := powMod(4, oddCount, MOD);

    return int((part1 * part2) % MOD);
}

func powMod(a, b, mod int64) int64 {
    result := int64(1);

    a %= mod;

    for b > 0 {
        if b % 2 == 1 {
            result = (result * a) % mod;
        }

        a = (a * a) % mod;
        b /= 2;
    }

    return result;
}

File: 2050-count-good-numbers/count-good-numbers.java
class Solution {
    private static final int MOD = (int) 1e9 + 7;

    public int countGoodNumbers(long n) {
        long evenCount = (n + 1) / 2;
        long oddCount = n / 2;

        long part1 = powMod(5, evenCount);
        long part2 = powMod(4, oddCount);

        return (int) ((part1 * part2) % MOD);
    }

    private long powMod(long a, long b) {
        long result = 1;

        a %= MOD;

        while (b > 0) {
            if ((b & 1) == 1) result = (result * a) % MOD;
            a = (a * a) % MOD;
            b >>= 1;
        }

        return result;
    }
}